validator_client/beacon_node_fallback/src/beacon_head_monitor.rs

```diff
@@ -0,0 +1,147 @@
+use crate::BeaconNodeFallback;
+use std::collections::HashMap;
+use tokio::sync::RwLock;
+
+use types::EthSpec;
+
+use eth2::types::{EventKind, EventTopic, SseHead};
+use tracing::{info, warn};
+
+use futures::StreamExt;
+use slot_clock::SlotClock;
+use std::sync::Arc;
+
+type CacheHashMap = HashMap<usize, SseHead>;
+
+// This is used send the index derived from `CandidateBeaconNode` to the
+// `AttestationService` for further processing
+#[derive(Debug)]
+pub struct HeadEvent {
+    pub beacon_node_index: usize,
```

Comment:
I think we could use a unique `id` to identify each BN, rather than its index. The index seems too flaky and liable to break.

e.g. at the moment the function `update_all_candidates` will reorder the nodes, but not clear the cache, which will break the link between head events and BNs.

To assign IDs we could keep track of a field `next_beacon_node_id: AtomicUsize` in `BeaconNodeFallback`, and use this to assign IDs to candidate beacon nodes whenever they are _added_. It should be immutable after `CandidateBeaconNode` creation. New IDs can be assigned by just doing `next_beacon_node_id.fetch_add(1, Ordering::SeqCst)`.